server/src/controllers/eventController.service.ts

```diff
@@ -0,0 +1,309 @@
+                  ticketrestrictionsid: restriction.ticketrestrictionsid,
+                },
+                data: {
+                  ticketssold:
+                  restriction.ticketlimit -
+                  (
+                    instance.ticketTypeMap.get(restriction.tickettypeid_fk) ??
+                    []
+                  ).length,
+                },
+              })),
+            },
+          },
+        }),
+    );
+  }
+  return {
+    cartRows,
+    orderItems,
+    orderTotal,
+    eventInstanceQueries,
+  };
+};
+
+const getTickets = (
+    prisma: PrismaClient,
+    availableTickets: Map<number, number[]>,
+    typeID: number,
+    quantity: number,
+    price: number,
+) => {
+  const ticketsForType = availableTickets.get(typeID) ?? [];
+  const ticketsForGA = availableTickets.get(1) ?? [];
+  if (
+    ticketsForType.length < quantity ||
+    (typeID !== 1 && ticketsForGA.length < quantity)
+  ) {
+    throw new InvalidInputError(422, `Requested Tickets no longer available`);
+  }
+
+  const ticketsToSellForType = ticketsForType.splice(0, quantity);
+  const ticketsToSellForGA =
+    typeID != 1 ? ticketsForGA.splice(0, quantity) : [];
+
+  return ticketsToSellForType.map((ticketID, index) => ({
+    price,
+    singletickets: {
+      create: {
+        eventtickets: {
+          connect: [
+            {
+              eventticketid: ticketID,
+            },
+          ].concat(
+            typeID != 1 ? [{eventticketid: ticketsToSellForGA[index]}] : [],
+          ),
+        },
+      },
+    },
+  }));
+};
+
+interface checkOutForm {
+  firstName: string;
+  lastName: string;
+  streetAddress: string;
+  postalCode: string;
+  country: string;
+  phone: string;
+  email: string;
+  visitSource: string;
+  seatingAcc: string;
+  comments: string;
+  optIn: boolean;
+}
+
+export const updateContact = async (
+    formData: checkOutForm,
+    prisma: PrismaClient,
+) => {
+  const existingContact = await prisma.contacts.findFirst({
+    where: {
+      email: formData.email,
+    },
+    select: {
+      contactid: true,
+    },
+  });
+
+  let updatedContact: { contactid: number } | null;
+  if (!existingContact) {
+    updatedContact = await prisma.contacts.create({
+      data: {
+        ...validateContact(formData),
+      },
+      select: {
+        contactid: true,
+      },
+    });
+  } else {
+    updatedContact = await prisma.contacts.update({
+      where: {
+        contactid: existingContact.contactid,
+      },
+      data: {
+        ...validateContact(formData),
+      },
+      select: {
+        contactid: true,
+      },
+    });
+  }
+  return updatedContact;
+};
+
+const validateContact = (formData: checkOutForm) => {
+  return {
+    firstname: validateName(formData.firstName, 'First Name'),
+    lastname: validateName(formData.lastName, 'Last Name'),
+    email: validateWithRegex(
+        formData.email,
+        `Email: ${formData.email} is invalid`,
+        new RegExp('.*'),
+    ),
```

Comment:
could use a more strict RegExp for email since this is the data used as the unique ID for tickets

really basic example
new RegExp('^.+@.+\\..+$')

currently you can submit the form even if its a bad email like this
![image](https://github.com/WonderTix/WonderTix/assets/35935079/c7860a6a-e725-44f6-8ecf-fab6d8aa11a4)